packages/core/src/typings/openapi.ts

```diff
@@ -202,46 +230,61 @@ export type Oas3ParameterStyle =
 
-export interface Oas3RequestBody {
+export interface Oas3RequestBody<T extends Oas3Schema | Oas3_1Schema = Oas3Schema | Oas3_1Schema> {
   description?: string;
   required?: boolean;
-  content: { [mime: string]: Oas3MediaType };
+  content: { [mime: string]: Oas3MediaType<T> };
 }
 
-export interface Oas3Responses {
-  [code: string]: Oas3Response;
+export interface Oas3Responses<T extends Oas3Schema | Oas3_1Schema> {
+  [code: string]: Oas3Response<T>;
 }
 
-export interface Oas3Response {
+export interface Oas3Response<T extends Oas3Schema | Oas3_1Schema = Oas3Schema | Oas3_1Schema> {
   description?: string;
-  headers?: { [name: string]: Referenced<Oas3Header> };
-  content?: { [mime: string]: Oas3MediaType };
+  headers?: { [name: string]: Referenced<Oas3Header<T>> };
+  content?: { [mime: string]: Oas3MediaType<T> };
   links?: { [name: string]: Referenced<Oas3Link> };
 }
 
 export interface Oas3Link {
-  $ref?: string;
+  operationRef?: string;
+  operationId?: string;
+  parameters?: { [name: string]: unknown };
+  requestBody?: unknown;
+  description?: string;
+  server?: Oas3Server;
 }
 
-export type Oas3Header = Omit<Oas3Parameter, 'in' | 'name'>;
+export type Oas3Header<T extends Oas3Schema | Oas3_1Schema = Oas3Schema | Oas3_1Schema> = Omit<
+  Oas3Parameter<T>,
+  'in' | 'name'
+>;
 
-export interface Oas3Callback {
-  [name: string]: Oas3PathItem;
+export interface Oas3Callback<T extends Oas3Schema | Oas3_1Schema = Oas3Schema | Oas3_1Schema> {
+  [name: string]: Oas3PathItem<T>;
 }
 
-export interface Oas3Components {
-  schemas?: { [name: string]: Referenced<Oas3Schema> };
-  responses?: { [name: string]: Referenced<Oas3Response> };
-  parameters?: { [name: string]: Referenced<Oas3Parameter> };
+// common fields for OAS components v3.x
+export interface Oas3ComponentsBase<T extends Oas3Schema | Oas3_1Schema> {
+  schemas?: { [name: string]: Referenced<T> };
+  responses?: { [name: string]: Referenced<Oas3Response<T>> };
+  parameters?: { [name: string]: Referenced<Oas3Parameter<T>> };
   examples?: { [name: string]: Referenced<Oas3Example> };
-  requestBodies?: { [name: string]: Referenced<Oas3RequestBody> };
-  headers?: { [name: string]: Referenced<Oas3Header> };
+  requestBodies?: { [name: string]: Referenced<Oas3RequestBody<T>> };
+  headers?: { [name: string]: Referenced<Oas3Header<T>> };
   securitySchemes?: { [name: string]: Referenced<Oas3SecurityScheme> };
   links?: { [name: string]: Referenced<Oas3Link> };
-  callbacks?: { [name: string]: Referenced<Oas3Callback> };
+  callbacks?: { [name: string]: Referenced<Oas3Callback<T>> };
 }
 
-export type Oas3ComponentName = keyof Oas3Components;
+export interface Oas3_1Components extends Oas3ComponentsBase<Oas3_1Schema> {
+  pathItems?: { [name: string]: Referenced<Oas3PathItem<Oas3_1Schema>> };
+}
+
+export interface Oas3Components extends Oas3ComponentsBase<Oas3Schema> {}
+
+export type Oas3ComponentName<T extends Oas3Schema | Oas3_1Schema> = keyof Oas3ComponentsBase<T>;
```

Comment:
Let's add a default for backward compatibility:

```suggestion
export type Oas3ComponentName<T extends Oas3Schema | Oas3_1Schema = Oas3Schema | Oas3_1Schema> = keyof Oas3ComponentsBase<T>;
```